reference/extensions/hooks.rst

```diff
@@ -1,4 +1,167 @@
+            if "git clone" in conanfile_content and "git checkout" in conanfile_content:
+                fixed_sources = True
+                for invalid in invalid_content:
+                    if invalid in conanfile_content:
+                        fixed_sources = False
+            else:
+                fixed_sources = False
+                for valid in valid_content:
+                    if valid in conanfile_content:
+                        fixed_sources = True
+
+            if not fixed_sources:
+                output.error(f"[IMMUTABLE SOURCES] Source files does not come from and immutable place. Checkout to a "
+                            "commit/tag or download a compressed source file for {conanfile.ref}")
+
+This hook checks the recipe content prior to it being exported and prior to downloading the sources. Basically the
+``pre_export()`` function checks the attributes of the ``conanfile`` object to see if there is an URL, a license and a description and if missing,
+warns the user with a message through the ``conanfile.output``. This is done **before** the recipe is exported to the local cache.
+
+The ``pre_source()`` function checks if the recipe contains a ``source()`` method (this time it is using the *conanfile.py* content instead of
+the ``conanfile`` object) and in that case it checks if the download of the sources are likely coming from immutable places (a compressed
+file or a determined :command:`git checkout`). This is done **before** the **source()** method of the recipe is called.
+
+Any kind of Python script can be executed. You can create global functions and call them from different hook functions, import from a
+relative module and warn, error or even raise to abort the Conan client execution.
+
+
+Other useful task where a hook may come handy are the upload and download actions. There are **pre** and **post** functions for every
+download/upload as a whole and for fine download tasks such as recipe and package downloads/uploads.
+
+For example they can be used to sign the packages (including a file with the signature) when the package is created and check that
+signature every time they are downloaded.
+
+.. code-block:: python
+   :caption: *hook_package_signing.py*
+
+    import os
+    from conan.tools.files import save, load
+
+    SIGNATURE = os.getenv("ENTERPRISE_CONAN_PKG_SIGNATURE", "default-signature")
+
+    def post_package(conanfile):
+        sign_path = os.path.join(conanfile.package_folder, ".sign")
+        save(conanfile, sign_path, SIGNATURE)
+        conanfile.output.info("Package signed successfully")
+
+    def post_package_info(conanfile):
+        sign_path = os.path.join(conanfile.package_folder, ".sign")
+        content = load(conanfile, sign_path)
+        if content != SIGNATURE:
+            conanfile.output.error(f"[CHECK SIGNATURE] The signature file '{sign_path}' has a different content than the expected one.")
```

Comment:
Same, this seems example, not reference.
Furthermore, the signing example is not good anymore that we have a dedicated plugin for signing packages.

Reply:
Removed the second example.